File: cse_241_object_oriented_programming/HW04/Game Of Life.cpp
/**********************************************************/
// Emrah YILDIRIM 111044056
// HW04_111044056.cpp


#include"file.h"


int Cell::takeCoor(void)
{
	int coor;
	
	cin >> coor;
	
	return coor;
}


void Cell::printCell(void)	const
{
	cout << getX() << "." << getY() <<endl;
}


inline int  Cell::getX(void)	const
{
	return x;
}


inline int  Cell::getY(void)	const
{
	return y;
}


inline void Cell::setX(const int& cX)
{
	x =cX;
}


inline void Cell::setY(const int& cY)
{
	x =cY;
}


void Cell::setCell(const int& cX, const int& cY)
{
	x = cX;
	y = cY;
}


/*****************************************************************************/



int GameOfLife::numOfAlive = 0;



void GameOfLife::setRow(const int& row)
{
	if(row >= 0)
		sizeRow = row;
	else
	{
		cout << "wrong row index\n";
		
		exit(1);
	}
}




void GameOfLife::setCol(const int& col)
{
	if(col >= 0)
		sizeCol = col;
	else
	{
		cout << "wrong column index\n";
		exit(1);
	}
}


inline int GameOfLife::getSizeRow(void) const
{
	return sizeRow;
}

inline int GameOfLife::getSizeCol(void) const
{
	return sizeCol;
}



void GameOfLife::readFile(const char *fileName)
{
	FILE *map;
	char cell, trash;
	
	map = fopen(fileName, "r");
	
	
	
	fscanf(map, "%d%d%c", &sizeRow, &sizeCol, &trash);
	
	
	for(int i=0; i < sizeRow; ++i)
	{
		for(int j=0; j < sizeCol; ++j)
		{
			fscanf(map, "%c", &cell);
			
			if(cell == 'X')
				livingsReal.push_back(Cell(i, j));
				// tek 1 obje oluþturmayý da dene 
				
		}
	
		fscanf(map, "%c", &trash);
	}	
	
	//eleman sayisi artirilir
	numOfAlive += livingsReal.size();
	fclose(map);
	
}



void GameOfLife::write2File(const char *fileName)
{
	FILE *outFile;
	
	outFile = fopen(fileName, "w");

	
	for(int i=0; i < sizeRow; ++i)
	{	
		for(int j=0; j < sizeCol; ++j)
		{
			if(isAlive(Cell(i,j)))
				fprintf(outFile, "X");
			else
				fprintf(outFile, ".");
		}
		fprintf(outFile, "\n");
	}

	fclose(outFile);
}


void GameOfLife::printVector()	const
{
	int size = livingsReal.size();

	for(int i=0; i < size; ++i)
		livingsReal[i].printCell();
		

}



void GameOfLife::printBoard()
{
	for(int i=0; i < sizeRow; ++i)
	{	
		for(int j=0; j < sizeCol; ++j)
		{
			if(isAlive(Cell(i,j)))
				cout<<"X";
			else
				cout<<".";
		}
		cout<<endl;
	}
}


int GameOfLife::findMaxRow()
{
	int size = livingsReal.size(), max=0, num=0;
	
	for(int i=0; i < size; ++i)
	{
		if((num=livingsReal[i].getX()) > max)
			max=num;
	}
	return max;
}



int GameOfLife::findMaxCol()
{
	int size = livingsReal.size(), max=0, num=0;
	
	for(int i=0; i < size; ++i)
	{ 
		if((num=livingsReal[i].getY()) > max)
			max=num;
	}
	return max;
}


void GameOfLife::play(void)
{
	
	syncVectors();
	printBoard();

	for(int i=0; i < sizeRow; ++i)
	{	
		for(int j=0; j < sizeCol; ++j)
			bornOrKill(Cell(i,j));	
	}
	
	if( findMaxRow() >= sizeRow-1 || findMaxCol() >= sizeCol-1)
	{
	   sizeRow++;
		sizeCol++;
	}
	cout <<endl;
	syncVectors();
	

}



int GameOfLife::bornOrKill(const Cell& cell)
{
	int status = howManyNB(cell);
	
	//canliysa olecek mi diye bakar
	//yoksa da canlanacak mi
	if(isAlive(cell))
	{	
		if(status < 2 || status > 3)
			killCell(cell);
	}
	else
	{
		if(status == 3)
		{
			livingsReal.push_back(cell);
			numOfAlive++;
		}
	}

}




int GameOfLife::howManyNB(const Cell& cell)
{
	int counter=0;
	int currentRow = cell.getX(), currentCol = cell.getY();
	
	
	if(isAlive(Cell(currentRow-1, currentCol-1)))
		counter++;
	if(isAlive(Cell(currentRow-1, currentCol)))
		counter++;
	if(isAlive(Cell(currentRow-1, currentCol+1)))
		counter++;
	if(isAlive(Cell(currentRow, currentCol-1)))
		counter++;
	if(isAlive(Cell(currentRow, currentCol+1)))
		counter++;
	if(isAlive(Cell(currentRow+1, currentCol-1)))
		counter++;
	if(isAlive(Cell(currentRow+1, currentCol)))
		counter++;
	if(isAlive(Cell(currentRow+1, currentCol+1)))
		counter++;
	
	return counter;
}



bool GameOfLife::isAlive(const Cell& cell)
{
	int size = livingsControl.size();
	
	
	for(int i=0; i < size; ++i)
	{
		if(livingsControl[i].getX() == cell.getX() &&
		   livingsControl[i].getY() == cell.getY())
			return true;
	}		
	return false;
}





bool GameOfLife::killCell(const Cell& cell)
{
	vector<Cell>::iterator point;
	int size=static_cast<int>(livingsReal.size());
	
	//oldurme yapilir eleman sayisi  azaltilir
	for(int i=0; i < size; ++i)
		if(livingsReal[i].getX() == cell.getX() && livingsReal[i].getY() == cell.getY())
		{
			point = livingsReal.begin() + i;
			livingsReal.erase(point);
			
			numOfAlive--;
			return true;
		}
	return false;
}



void GameOfLife::mergeGames(GameOfLife& other)
{
	int otherSize= other.livingsReal.size();
	
	// hangi objenin board u daha buyukse ona gecilir
	//satir ve sutun ayri ayi kontrol edilir
	if(other.findMaxRow() > findMaxRow())
		setRow(other.getSizeRow());
	
	if(other.findMaxCol() > findMaxCol())
		setCol(other.getSizeCol());
		
	
	for(int i=0; i < otherSize; ++i)
	{
		livingsReal.push_back(other.livingsReal[i]);
	
	}
}



void GameOfLife::syncVectors()
{
	livingsControl = livingsReal;
}

File: cse_241_object_oriented_programming/HW09/GameOfLife.java
/**
 * @author 111044056 Emrah YILDIRIM
 * GameOfLife.java from HW09
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.Scanner;

public class GameOfLife {
	
	private Cell [] livingsReal;
	private Cell [] livingsControl;	
	private static int numOfAlive = 0;
	private int sizeRow=0, sizeCol=0;
	private int counterControl=0, counter=0, capacity=20;
	
	GameOfLife() 
	{ 
		livingsReal = new Cell[capacity];	
	}
	
	/**
	 * find how many cells in all games
	 * @return all cells number
	 */
	public static int getAllObjNum(){ return numOfAlive; }
	
	/**
	 * @return row size
	 */
	public int getRowSize() { return sizeRow; }
	
	/**
	 * @return all column size
	 */
	public int getColSize() { return sizeCol; }
	
	/**
	 * @param row number
	 */
	public void setRow( int row){ sizeRow = row; }
	
	/**
	 * @param col cloumn number
	 */
	public void setCol( int col) { sizeCol = col; }
	
		
	/**
	 * find max row number in game
	 * @return max row number
	 */
	public int findMaxRow()
	{
		int max=0, num=0;
		
		for(int i=0; i < counter; ++i)
		{
			if((num=livingsReal[i].getX()) > max)
				max=num;
		}
		return max;
	}
	
	
	/**
	 * find max column number in game
	 * @return column row number
	 */
	public int findMaxCol()
	{
		int max=0, num=0;
		
		for(int i=0; i < counter; ++i)
		{
			if((num=livingsReal[i].getY()) > max)
				max=num;
		}
		return max;
	}
	
	
	
	/**
	 * find the cell if is alive
	 * @param cell cell object for control
	 * @return is alive or not 
	 */
	public boolean isAlive(Cell cell)
	{
		
		for(int i=0; i < counterControl; ++i)
		{
			if(livingsControl[i].getX() == cell.getX() &&
			   livingsControl[i].getY() == cell.getY())
				return true;
		}		
		return false;
	}

	
	/**
	 * print the board on screen
	 */
	public void printBoard()
	{
		int startRow=0, startCol=0;
		
		for(int i = 0; i < sizeCol; ++i)
			if(isAlive(new Cell(-1, i)))
			{
				startRow = -1;
				break;
			}
				
		for(int i = 0; i < sizeRow; ++i)
			if(isAlive(new Cell(i, -1)))
			{
				startCol = -1;
				break;
			}
				
		for(int i=startRow; i < sizeRow; ++i)
		{	
			for(int j=startCol; j < sizeCol; ++j)
			{
				if(isAlive(new Cell(i,j)))
					System.out.printf("X");
				else
					System.out.printf(".");
			}
			System.out.printf("\n");
		}
		System.out.printf("\n");
	}

	

	/**
	 * do life job for each cell parameter
	 * @param cell cell object for life job
	 */
	public void bornOrKill(Cell cell)
	{
		int status = howManyNB(cell);
		
		//canliysa olecek mi diye bakar
		//yoksa da canlanacak mi
		if(isAlive(cell))
		{	
			if(status < 2 || status > 3)
				killCell(cell);
		}
		else
		{
			if(status == 3)
				addElement(cell);

		}
	}

	
	/**
	 * count neighbor number
	 * @param cell cell object for counting
	 * @return neighbor count
	 */
	public int howManyNB(Cell cell)
	{
		int countNB=0;
 		int currentRow = cell.getX(), currentCol = cell.getY();
		
		
		if(isAlive(new Cell(currentRow-1, currentCol-1)))
			countNB++;
		if(isAlive(new Cell(currentRow-1, currentCol)))
			countNB++;
		if(isAlive(new Cell(currentRow-1, currentCol+1)))
			countNB++;
		if(isAlive(new Cell(currentRow, currentCol-1)))
			countNB++;
		if(isAlive(new Cell(currentRow, currentCol+1)))
			countNB++;
		if(isAlive(new Cell(currentRow+1, currentCol-1)))
			countNB++;
		if(isAlive(new Cell(currentRow+1, currentCol)))
			countNB++;
		if(isAlive(new Cell(currentRow+1, currentCol+1)))
			countNB++;
		
		//System.out.printf("%d %b\n", countNB, isAlive(new Cell(currentRow+1, currentCol+1)));
		return countNB;
	}

	

	/**
	 * delete the cell that was passed paramater
	 * @param cell object for kill job
	 */
	public void killCell(Cell cell)
	{		
		Cell [] temp = new Cell[capacity];
		int j=0;
		
		for(int i=0; i < counter; ++i)
			if(livingsReal[i].getX() == cell.getX() && livingsReal[i].getY() == cell.getY())
				continue;
			else
			{
				temp[j] = (Cell) livingsReal[i].clone();
				++j;
			}
			
		livingsReal = temp;
		//temp=livingsControl;
		counter--;
		numOfAlive--;
				
	}

	
	
/**
 * merge two games
 * @param other game
 */
	public void mergeGames(GameOfLife other)
	{
		int otherSize = other.counter;
		boolean flag=false;
		
		// hangi objenin board u daha buyukse ona gecilir
		// satir ve sutun ayri ayri kontrol edilir
		if(other.findMaxRow() > findMaxRow())
			setRow(other.getRowSize());
		
		if(other.findMaxCol() > findMaxCol())
			setCol(other.getColSize());
			
		
		for(int i=0; i < otherSize; ++i)
		{
			if(!isAlive(other.livingsReal[i]))
			{
				addElement(other.livingsReal[i]);
				flag = true;
			}
		
		}

		if(flag)
			sync();
		
	}

	
	/**
	 * sync two arrays
	 */
	public void sync()
	{
		livingsControl = livingsReal.clone();
		counterControl = counter;
	}
	

	/**
	 * play the game one step
	 */
	public void play()
	{
		sync();
		
		if( findMaxRow() >= sizeRow-1 || findMaxCol() >= sizeCol-1)
		{
		   sizeRow++;
		   sizeCol++;
		}
		
		for(int i=-1; i < sizeRow; ++i)
		{	
			for(int j=-1 ; j < sizeCol; ++j)
				bornOrKill(new Cell(i,j));	
		}
		
		sync();
		

		printBoard();
		System.out.println("\n");
		
	
	}
	
	/**
	 * add cell parameter to Cell array
	 * @param cell object for adding
	 */
	public void addElement(Cell cell)
	{
		
		if(counter < capacity)
		{
			livingsReal[counter] = new Cell(cell.getX(), cell.getY());
			counter++;
			numOfAlive++;
		}
		else
		{
			capacity *= 2;
			Cell [] temp = new Cell[capacity];
			Arrays.fill(temp, new Cell(-1,-1));
			for(int i=0; i < capacity/2; ++i)
			{
				temp[i].setX(livingsReal[i].getX());
				temp[i].setY(livingsReal[i].getY());
			}
			livingsReal = temp;
			
			livingsReal[counter].setX(cell.getX());
			livingsReal[counter].setY(cell.getY());
			counter++;
			numOfAlive++;
			
		}
	
	}
	
	
	/**
	 * read from file the maps
	 * @param fileName
	 * @throws FileNotFoundException
	 */

	void readFile(String fileName) throws FileNotFoundException
	{
		/*File map = new File(fileName);
		FileReader readFile = new FileReader(map);
		BufferedReader readInteger = new BufferedReader(readFile);
		
		readInteger.read*/
		int cell;
		Scanner map = new Scanner(new File(fileName));
		
		sizeRow = map.nextInt();
		sizeCol = map.nextInt();	
		
			
		for(int i=0; i < sizeRow; ++i)
		{
			for(int j=0; j < sizeCol; ++j)
			{
				cell = map.nextInt();
				if(cell == 1)				
					addElement(new Cell(i, j));
			}
		}	
	
	sync();
	
	
	map.close();
}

	
	
}
